### PR TITLE
[KOGITO-146] Implement CI for build and push Operator Image to Quay

### PR DESCRIPTION
**Pull request by sbuvaneshkumar**
There are few variables that needs to be set in circle CI project settings, in order for this configuration to work.
1. CIRCLE_TAG - release tag 
2. REGISTRY_USER - registry.redhat.io username
3. REGISTRY_PASS - registry password
4. QUAY_USER - quay username
5. QUAY_PASS - quay password
6. TOKEN - Quay token (without basic)

release tags should not be start with v (such as v0.3.0).

Sample test results:
     log: https://circleci.com/gh/sbuvaneshkumar/kogito-cloud-operator/33 
     repo: https://quay.io/repository/sbuvaneshkumar/kogito-cloud-operator?tab=tags

Reference: 
https://github.com/aerogear/mobile-security-service-operator/blob/master/.circleci/config.yml
https://github.com/aerogear/mobile-security-service-operator#tags-release
https://github.com/aerogear/mobile-security-service-operator/blob/master/Makefile

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -15,3 +15,62 @@ jobs:
...
+  image_release:
+    working_directory: /go/src/github.com/kiegroup/kogito-cloud-operator
+    docker:
+      - image: circleci/golang:1.12
+        environment: GO111MODULE=on
+    steps:
+      - checkout
+      - setup_remote_docker
+      - run:
+          name: Install dep
```
Comment (ricardozanini) [inline .circleci/config.yml]: We don't use dep anymore, take a look at go-mod.sh

File: .circleci/config.yml [hunk 2]
```diff
@@ -15,3 +15,62 @@ jobs:
       - run: go fmt ./...
       - run: go vet ./...
       - run: go test -mod=vendor ./pkg/... ./cmd/... -count=1
+  image_push_master:
```
Comment (ricardozanini) [inline .circleci/config.yml]: Besides pushing the image, we should push the `application` to quay as well, see: https://github.com/kiegroup/kogito-cloud-operator#deploy-to-openshift-4x-for-development-purposes

**Comment by spolti** [comment]
I like the idea to automatically push images to quay, but I have some concerns about it;.
Which account are you using?
Where the credentials are stored?
If for some reason someone finds it they can delete our images.

IMHO steps like these must be triggered manually by our internal infra.

**Comment by sbuvaneshkumar** [comment]
> I like the idea to automatically push images to quay, but I have some concerns about it;.
> Which account are you using?
CircleCI account? I believe the account for this is kiegroup. 

> Where the credentials are stored?
Credentials will be stored in Circle CI project settings. It is a secure place and not everyone can access it, apart from the owner who owns the circleci account for this project.

> If for some reason someone finds it they can delete our images.
I don't think anyone can delete our images, since we own the (quay) account.

> IMHO steps like these must be triggered manually by our internal infra.

File: .circleci/config.yml (hunk 2 above)
Comment (sbuvaneshkumar) [reply .circleci/config.yml]: @ricardozanini Thanks for the review :). I will modify the PR accordingly.

**Comment by sbuvaneshkumar** [comment]
@ricardozanini Currently the images are generated and tagged with master when any change happens in master branch and then pushed to Quay. 
When tags specified, then the image is pushed to Quay tagged with the actual tag.

I believe we need the same behaviour for pushing applications as well, if yes and then I will go ahead and make the workflow changes in circleci config accordingly. 
Please let me know if any change required in this process.

Note: I have used my circleci/Quay accounts, after finalizing the process, I will change that back to kogito. Also, someone who have access to kogito CircleCI project settings needs to add the variables mentioned above.

https://circleci.com/gh/sbuvaneshkumar/workflows/kogito-cloud-operator/tree/master

**Comment by ricardozanini** [comment]
@sbuvaneshkumar many thanks putting so much efforts on it. I'll take a look at the changes carefully later.

@spolti do you mind taking a look also when you have the time?

**Comment by ricardozanini** [comment]
Hi @sbuvaneshkumar, some things to consider:

1. I'd say that we could build and push the image to quay each time a PR is merged to master with the following tag: `quay.io/kiegroup/kogito-cloud-operator:version-RC`, this way we will have a brand new version of the operator to work with
2. Forget about the application for now, this will be another CI to release the operator
3. We should use an application credential from quay that only has the permissions to write in this specific repository (kogito-cloud-operator). Could you add this info to the docs somehow (like a CI doc in the `docs` folder)?
4. Rebase your branch and squash your commits after this change, please

@spolti @sutaakar  WDYT?

File: Makefile [hunk 3]
```diff
@@ -65,4 +75,39 @@ addheaders:
...
+.PHONY: image/build/release
+image/build/release:
+	@echo Building operator with the tag: $(IMAGE_RELEASE_TAG)
+	@docker login --username $(REGISTRY_USER) --password $(REGISTRY_PASS) https://registry.redhat.io
+	operator-sdk build $(IMAGE_RELEASE_TAG)
+	operator-sdk build $(IMAGE_LATEST_TAG)
+
+.PHONY: image/push/master
+image/push/master:
+	@echo Pushing operator with tag $(IMAGE_MASTER_TAG) to $(IMAGE_REGISTRY)
```
Comment (sutaakar) [inline Makefile]: IMAGE_REGISTRY is used here as parameter, but it is not used in commands below, push is hardcoded to quay.io

**Comment by sutaakar** [comment]
@ricardozanini Sounds good to me.

**Comment by sbuvaneshkumar** [comment]
> Hi @sbuvaneshkumar, some things to consider:
> 
> 1. I'd say that we could build and push the image to quay each time a PR is merged to master with the following tag: `quay.io/kiegroup/kogito-cloud-operator:version-RC`, this way we will have a brand new version of the operator to work with

Currently, the configs are meant to tag the image with "master" tag whenever a change happens in master branch. sure, I will change the tag to "version-RC".

> 2. Forget about the application for now, this will be another CI to release the operator

@ricardozanini  Can you please clarify if you mean we do not want to push the application to Quay as of now in this CI?

> 3. We should use an application credential from quay that only has the permissions to write in this specific repository (kogito-cloud-operator). Could you add this info to the docs somehow (like a CI doc in the `docs` folder)?

Ack.

> 4. Rebase your branch and squash your commits after this change, please

Ack.
> @spolti @sutaakar WDYT?

**Comment by ricardozanini** [comment]
> @ricardozanini Can you please clarify if you mean **we do not want to push the application to Quay** as of now in this CI?
That's exactly what I meant!

> Currently, the configs are meant to tag the image with "master" tag whenever a change happens in master branch. sure, I will change the tag to "version-RC".
This is the current version of the master, you can fix this in the circleci file or take it from here:
https://github.com/kiegroup/kogito-cloud-operator/blob/master/version/version.go#L19 (preferred)

**Comment by sbuvaneshkumar** [comment]
@spolti @ricardozanini I have added the changes as per the request.

**Comment by ricardozanini** [comment]
@sbuvaneshkumar I couldn't come up with an idea to have this CI run without our credentials exposed to a third party service. I'll close this PR and JIRA for now until we have a better solution or Quay.io have implemented the application tokens. Also, have to expose our internal Red Hat registry credentials is not a good idea either.

Many thanks for your contribution, I'll use it as a reference for when we come back to this later.

**Pull request closed without merging**